File: rust/src/messages.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MessageRole {
    System,
    Developer,
    User,
    Assistant,
    Tool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
    pub id: String,
    pub name: String,
    pub arguments: String,
}

#[derive(Debug, Clone)]
pub struct Message {
    pub role: MessageRole,
    pub content: String,
    pub tool_calls: Vec<ToolCall>,
    pub tool_call_id: Option<String>,
    pub name: Option<String>,
    pub reasoning_content: Option<String>,
    pub images: Vec<String>,
    /// Normalized Responses API content blocks when the caller supplied
    /// structured Chat Completions content. `None` retains the legacy
    /// `content` + `images` representation.
    pub structured_content: Option<Vec<serde_json::Value>>,
}

impl Message {
    pub fn new(
        role: MessageRole,
        content: String,
        tool_calls: Vec<ToolCall>,
        tool_call_id: Option<String>,
        name: Option<String>,
    ) -> Result<Self, MessageError> {
        if role == MessageRole::Tool {
            if tool_call_id.is_none() {
                return Err(MessageError::Validation(
                    "tool messages require a string tool_call_id".to_string(),
                ));
            }
            if name.as_deref().is_some_and(str::is_empty) {
                return Err(MessageError::Validation(
                    "tool message name must be non-empty when provided".to_string(),
                ));
            }
        } else if tool_call_id.is_some() || name.is_some() {
            return Err(MessageError::Validation(
                "tool_call_id and name are only allowed on tool messages".to_string(),
            ));
        }
        if !tool_calls.is_empty() && role != MessageRole::Assistant {
            return Err(MessageError::Validation(
                "tool_calls are only allowed on assistant messages".to_string(),
            ));
        }
        Ok(Self {
            role,
            content,
            tool_calls,
            tool_call_id,
            name,
            reasoning_content: None,
            images: vec![],
            structured_content: None,
        })
    }
}

#[derive(Debug, thiserror::Error)]
pub enum MessageError {
    #[error("{0}")]
    Validation(String),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Usage {
    pub prompt_tokens: i64,
    pub completion_tokens: i64,
    pub total_tokens: i64,
    pub cached_tokens: Option<i64>,
    pub cache_write_tokens: Option<i64>,
}

#[derive(Debug, Clone)]
pub struct AssistantResponse {
    pub content: String,
    pub tool_calls: Vec<ToolCall>,
    pub finish_reason: Option<String>,
    pub usage: Option<Usage>,
    pub reasoning_content: Option<String>,
    pub raw: Option<serde_json::Value>,
    pub response_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolSchema {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub parameters: serde_json::Value,
    pub strict: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_message_user() {
        let msg = Message::new(MessageRole::User, "hello".to_string(), vec![], None, None).unwrap();
        assert_eq!(msg.role, MessageRole::User);
        assert_eq!(msg.content, "hello");
    }

    #[test]
    fn test_message_tool_requires_fields() {
        let result = Message::new(MessageRole::Tool, "output".to_string(), vec![], None, None);
        assert!(result.is_err());
    }

    #[test]
    fn test_message_tool_valid() {
        let msg = Message::new(
            MessageRole::Tool,
            "output".to_string(),
            vec![],
            Some("call-1".to_string()),
            Some("my_tool".to_string()),
        )
        .unwrap();
        assert_eq!(msg.role, MessageRole::Tool);
    }

    #[test]
    fn test_message_tool_name_is_optional() {
        let msg = Message::new(
            MessageRole::Tool,
            "output".to_string(),
            vec![],
            Some("call-1".to_string()),
            None,
        )
        .unwrap();
        assert_eq!(msg.tool_call_id.as_deref(), Some("call-1"));
        assert!(msg.name.is_none());
    }

    #[test]
    fn test_message_non_tool_with_tool_call_id() {
        let result = Message::new(
            MessageRole::User,
            "hi".to_string(),
            vec![],
            Some("call-1".to_string()),
            None,
        );
        assert!(result.is_err());
    }

    #[test]
    fn test_message_tool_calls_on_non_assistant() {
        let tc = ToolCall {
            id: "c1".to_string(),
            name: "fn".to_string(),
            arguments: "{}".to_string(),
        };
        let result = Message::new(MessageRole::User, "hi".to_string(), vec![tc], None, None);
        assert!(result.is_err());
    }

    #[test]
    fn test_message_assistant_with_tool_calls() {
        let tc = ToolCall {
            id: "c1".to_string(),
            name: "fn".to_string(),
            arguments: "{}".to_string(),
        };
        let msg =
            Message::new(MessageRole::Assistant, "".to_string(), vec![tc], None, None).unwrap();
        assert_eq!(msg.tool_calls.len(), 1);
    }

    #[test]
    fn usage_preserves_explicit_total() {
        let u = Usage {
            prompt_tokens: 100,
            completion_tokens: 50,
            total_tokens: 150,
            cached_tokens: Some(20),
            cache_write_tokens: None,
        };
        assert_eq!(u.total_tokens, 150);
        assert_eq!(u.cached_tokens, Some(20));
    }

    #[test]
    fn usage_preserves_cache_write_tokens() {
        let u = Usage {
            prompt_tokens: 100,
            completion_tokens: 50,
            total_tokens: 150,
            cached_tokens: Some(0),
            cache_write_tokens: Some(9),
        };
        assert_eq!(u.cache_write_tokens, Some(9));
    }

    #[test]
    fn usage_serialization_keeps_actual_counts() {
        let u = Usage {
            prompt_tokens: 4,
            completion_tokens: 3,
            total_tokens: 7,
            cached_tokens: Some(2),
            cache_write_tokens: None,
        };
        let value = serde_json::to_value(u).unwrap();
        assert_eq!(value["prompt_tokens"], 4);
        assert_eq!(value["completion_tokens"], 3);
        assert_eq!(value["total_tokens"], 7);
    }

    #[test]
    fn usage_deserialization_requires_all_required_counts() {
        assert!(serde_json::from_value::<Usage>(serde_json::json!({
            "prompt_tokens": 1,
            "completion_tokens": 1,
            "cached_tokens": 0,
            "cache_write_tokens": null
        }))
        .is_err());
    }
}

File: src/codex_as_api/messages.py
from __future__ import annotations

import dataclasses
import enum


class MessageRole(enum.Enum):
    SYSTEM = "system"
    DEVELOPER = "developer"
    USER = "user"
    ASSISTANT = "assistant"
    TOOL = "tool"


@dataclasses.dataclass(frozen=True, slots=True)
class ToolCall:
    id: str
    name: str
    arguments: str

    def __post_init__(self) -> None:
        if not isinstance(self.id, str):
            raise TypeError("ToolCall.id must be a string")
        if not isinstance(self.name, str):
            raise TypeError("ToolCall.name must be a string")
        if not isinstance(self.arguments, str):
            raise TypeError("ToolCall.arguments must be a string")


@dataclasses.dataclass(frozen=True, slots=True)
class Message:
    role: MessageRole
    content: str
    tool_calls: tuple[ToolCall, ...] = ()
    tool_call_id: str | None = None
    name: str | None = None
    reasoning_content: str | None = None
    images: tuple[str, ...] = ()
    content_parts: tuple[dict[str, object], ...] | None = None

    def __post_init__(self) -> None:
        if not isinstance(self.role, MessageRole):
            raise TypeError("Message.role must be a MessageRole")
        if not isinstance(self.content, str):
            raise TypeError("Message.content must be a string")
        if not isinstance(self.tool_calls, tuple):
            object.__setattr__(self, "tool_calls", tuple(self.tool_calls))
        if self.content_parts is not None and not isinstance(self.content_parts, tuple):
            object.__setattr__(self, "content_parts", tuple(self.content_parts))
        if not isinstance(self.images, tuple):
            object.__setattr__(self, "images", tuple(self.images))
        if any(not isinstance(image, str) or not image for image in self.images):
            raise ValueError("Message.images must contain non-empty strings")
        if self.content_parts is not None and any(not isinstance(part, dict) for part in self.content_parts):
            raise TypeError("Message.content_parts must contain dict values")
        if self.reasoning_content is not None and not isinstance(
            self.reasoning_content,
            str,
        ):
            raise TypeError("Message.reasoning_content must be a string or None")
        if self.role is MessageRole.TOOL:
            if not isinstance(self.tool_call_id, str):
                raise TypeError("tool messages require a string tool_call_id")
            if self.name is not None and (not isinstance(self.name, str) or not self.name):
                raise ValueError("tool message name must be non-empty when provided")
        elif self.tool_call_id is not None or self.name is not None:
            raise ValueError("tool_call_id and name are only allowed on tool messages")
        if self.tool_calls and self.role is not MessageRole.ASSISTANT:
            raise ValueError("tool_calls are only allowed on assistant messages")


@dataclasses.dataclass(frozen=True, slots=True)
class Usage:
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int
    cached_tokens: int | None = None
    cache_write_tokens: int | None = None

    def __post_init__(self) -> None:
        for field, value in (
            ("prompt_tokens", self.prompt_tokens),
            ("completion_tokens", self.completion_tokens),
        ):
            if not isinstance(value, int) or isinstance(value, bool) or value < 0:
                raise ValueError(f"Usage.{field} must be a non-negative integer")
        if self.cache_write_tokens is not None and (
            not isinstance(self.cache_write_tokens, int)
            or isinstance(self.cache_write_tokens, bool)
            or self.cache_write_tokens < 0
        ):
            raise ValueError("Usage.cache_write_tokens must be a non-negative integer or None")
        if self.cached_tokens is not None and (
            not isinstance(self.cached_tokens, int) or isinstance(self.cached_tokens, bool) or self.cached_tokens < 0
        ):
            raise ValueError("Usage.cached_tokens must be a non-negative integer or None")
        if not isinstance(self.total_tokens, int) or isinstance(self.total_tokens, bool) or self.total_tokens < 0:
            raise ValueError("Usage.total_tokens must be a non-negative integer")
        if self.total_tokens != self.prompt_tokens + self.completion_tokens:
            raise ValueError("Usage.total_tokens must equal prompt_tokens plus completion_tokens")


@dataclasses.dataclass(frozen=True, slots=True)
class AssistantResponse:
    content: str
    finish_reason: str | None
    tool_calls: tuple[ToolCall, ...] = ()
    usage: Usage | None = None
    reasoning_content: str | None = None
    raw: dict | None = None
    response_id: str | None = None

    def __post_init__(self) -> None:
        if not isinstance(self.content, str):
            raise TypeError("AssistantResponse.content must be a string")
        if self.finish_reason not in {None, "stop", "tool_calls"}:
            raise ValueError("AssistantResponse.finish_reason must be null, stop, or tool_calls")
        if self.response_id is not None and (not isinstance(self.response_id, str) or not self.response_id):
            raise ValueError("AssistantResponse.response_id must be non-empty when provided")
        if not isinstance(self.tool_calls, tuple):
            object.__setattr__(self, "tool_calls", tuple(self.tool_calls))


class InterruptIdleSignal(Exception):  # noqa: N818 - this is an internal control-flow signal.
    """Raised when an interrupted agent turn should return to idle."""


@dataclasses.dataclass(frozen=True, slots=True)
class TerminatorCall:
    name: str
    arguments: dict

    def __post_init__(self) -> None:
        if not self.name:
            raise ValueError("TerminatorCall.name must be non-empty")
        if not isinstance(self.arguments, dict):
            raise TypeError("TerminatorCall.arguments must be a dict")


@dataclasses.dataclass(frozen=True, slots=True)
class AgentResponse:
    text: str
    finish_reason: str
    terminator_call: TerminatorCall | None = None
    usage: Usage | None = None
    reasoning_content: str | None = None
    raw: dict | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class ToolResult:
    ok: bool
    content: str
    payload: dict | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class ToolSchema:
    name: str
    description: str | None
    parameters: dict
    strict: bool = False

    def __post_init__(self) -> None:
        if not isinstance(self.name, str) or not self.name:
            raise ValueError("ToolSchema.name must be a non-empty string")
        if self.description is not None and not isinstance(self.description, str):
            raise TypeError("ToolSchema.description must be a string or None")
        if not isinstance(self.parameters, dict):
            raise TypeError("ToolSchema.parameters must be a dict")
        if not isinstance(self.strict, bool):
            raise TypeError("ToolSchema.strict must be a boolean")
